### DISCUSSION TITLE
Switching to termwiz from crossterm

### DISCUSSION THREAD

**Discussion by A-Walrus** [helix-editor/helix#4332] [General]
@archseer has said that he is thinking of switching to using termwiz instead of crossterm: [comment](https://github.com/helix-editor/helix/issues/1472#issuecomment-1008644709)

Opening this discussion to discuss pros and cons, and course of action.
**Pros:**
- Support for [hyperlinks](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda), which will allow clicking on hyperlinks within documentation popups to take you to documentation site.
- Builtin  `Surface` and `Cells` which we currently implement ourselves and with better render diffing than what we currently have.
- Checking terminal capabilities
- Widgets, which we currently implement ourselves
- Uses `usize` instead of `u16` which will eliminate outright the bugs that I have recently been trying to solve: #4318, #4172
- I believe I heard somewhere that it has better input handling and will allow mapping some inputs that `crossterm` has issues with, but I'm not sure...

**Cons:**
 - A fair bit of code will need to be rewritten in order to make the switch.

If you have any other thoughts about `crossterm` vs `termwiz` please post here

**Comment by CptPotato** [comment]
Does termwiz support all the platforms that are currently supported by crossterm?

**Reply by A-Walrus** [reply]
I believe so. The docs say it supports `unix style ttys and Windows style console APIs.` which I take to mean that Linux, Mac and Windows should all be supported.

**Comment by David-Else** [comment]
Does https://github.com/helix-editor/helix/issues/3725 work out of the box in termwiz, or if not, does it support it?

**Comment by CptPotato** [comment]
I wonder if this will also help with some of the input issues with different keyboard layouts.

For example, I am unable to use some keybinds like `~`, `|` or any of the "unimpaired" (`[`/`]`) movements because of how modifier keys are recognized.

**Comment by dead10ck** [comment]
I've just come to learn that termwiz does not compile on Android due to unconditional use of `shm`, which is not supported on Android. This would be a pretty big deal for me, as I frequently edit with helix in Termux.

**Reply by yyogo** [reply]
looks like termwiz now compiles on android: https://github.com/wez/wezterm/pull/2694

The shm use is still there (it panics on Android instead), but looks like it's only used by Kitty image encoding/decoding code which probably isn't critical for Helix

**Comment by rizary** [comment]
Hello, I found this when researching between crossterm and termwiz. May I ask is this still an option? I'm starting to start my own tui project and wondering if you'll choose termwiz for a new project?

**Reply by kirawi** [reply]
The switch isn't valuable as crossterm largely has feature parity with termwiz now (in the areas we need it to at least).
